### PR TITLE
feat: surface descriptive session key errors

### PR DESCRIPTION
**Pull request by dphilipson**
The session key plugin throws opaque errors when permission checks fail. Surface better errors in the SDK after a validation failure by using state overrides to replace the session key code with a version that produces descriptive errors.

---

## PR-Codex overview
This PR updates the `sessionKeyPluginActions` in the `extension.ts` file to improve error handling and permissions checks.

### Detailed summary
- Added `executeWithSessionKey` function for descriptive errors
- Introduced `fixedExecuteWithSessionKey` for improved error handling
- Included `getRpcErrorMessageFromViemError` function
- Defined `SessionKeyPermissionError` class
- Updated `DEBUG_SESSION_KEY_BYTECODE` in a separate file

> The following files were skipped due to too many changes: `packages/accounts/src/msca/plugins/session-key/debug-session-key-bytecode.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by dphilipson** [comment]
This is a draft version: it uses `any` to suppress a hairy type error and includes a testing script, `testo.ts`, which will be deleted when ready to publish.

File: packages/accounts/package.json [hunk 1]
```diff
@@ -75,6 +75,7 @@
   "gitHead": "ee46e8bb857de3b631044fa70714ea706d9e317d",
   "dependencies": {
     "@alchemy/aa-core": "^3.18.2",
+    "@alchemy/aa-alchemy": "^3.18.2",
```
Comment (dphilipson) [inline packages/accounts/package.json:78]: This is here for local dev purposes. I'll remove it when moving this out of draft status.

File: packages/accounts/src/msca/plugins/session-key/extension.ts [hunk 2]
```diff
@@ -104,11 +111,82 @@ export const sessionKeyPluginActions: <
...
+      overrides,
+      context,
+      account = client.account,
+    } = originalArgs[0];
+    if (!account) {
+      throw initialError;
+    }
+    const data = og.encodeExecuteWithSessionKey({ args });
+    const sessionKeyPluginAddress =
+      SessionKeyPlugin.meta.addresses[client.chain.id];
```
Comment (dphilipson) [inline packages/accounts/src/msca/plugins/session-key/extension.ts:150]: As with other uses of this plugin address lookup, we should probably be using a default address if the chain id is unknown.

File: packages/accounts/src/msca/plugins/session-key/extension.ts [hunk 3]
```diff
@@ -104,11 +111,82 @@ export const sessionKeyPluginActions: <
...
+    try {
+      await client.buildUserOperation({
+        uo: data,
+        overrides: {
+          ...overrides,
+          stateOverride: {
+            ...overrides?.stateOverride,
+            [sessionKeyPluginAddress]: {
+              code: DEBUG_SESSION_KEY_BYTECODE,
+            },
```
Comment (dphilipson) [inline packages/accounts/src/msca/plugins/session-key/extension.ts:164]: The typings for state overrides are currently broken. The type system wants `stateOverride` to be an array of objects with `address` fields, but this is passed directly to the bundler rpc endpoints which expect an mapping of address -> override objects.

File: packages/accounts/src/msca/plugins/session-key/extension.ts [hunk 4]
```diff
@@ -104,11 +111,82 @@ export const sessionKeyPluginActions: <
...
+          },
+        },
+        context,
+        account,
+      } as any);
```
Comment (dphilipson) [inline packages/accounts/src/msca/plugins/session-key/extension.ts:169]: Even apart from the above type issue, I get type errors for _any_ nonempty `stateOverride` object, hence this `any` cast. Will need to debug.

File: packages/accounts/src/testo.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
+import { LocalAccountSigner, arbitrumSepolia } from "@alchemy/aa-core";
```
Comment (dphilipson) [inline packages/accounts/src/testo.ts:1]: Script used for local testing. You can try it yourself if you want! Will delete when moving out of draft status.

File: packages/accounts/src/msca/plugins/session-key/extension.ts [hunk 6]
```diff
@@ -104,11 +111,84 @@ export const sessionKeyPluginActions: <
...
+    );
+    const updatedOverrides: UserOperationOverrides<TEntryPointVersion> = {
+      ...(overrides as UserOperationOverrides<TEntryPointVersion>),
+      stateOverride: [
+        ...(overrides?.stateOverride ?? []),
+        {
+          address: sessionKeyPluginAddress,
+          code: DEBUG_SESSION_KEY_BYTECODE,
+        },
+      ],
```
Comment (dphilipson) [inline packages/accounts/src/msca/plugins/session-key/extension.ts:163]: This use of state overrides is currently broken and depends on the fix in https://github.com/alchemyplatform/aa-sdk/pull/720.

File: packages/accounts/src/msca/plugins/session-key/extension.ts [hunk 7]
```diff
@@ -104,11 +111,84 @@ export const sessionKeyPluginActions: <
...
+    }
+    const data = og.encodeExecuteWithSessionKey({ args });
+    const sessionKeyPluginAddress =
+      SessionKeyPlugin.meta.addresses[client.chain.id];
+    // The bytecode is 10kb of hex. Only import it if needed.
+    const { DEBUG_SESSION_KEY_BYTECODE } = await import(
+      "./debug-session-key-bytecode.js"
+    );
+    const updatedOverrides: UserOperationOverrides<TEntryPointVersion> = {
+      ...(overrides as UserOperationOverrides<TEntryPointVersion>),
```
Comment (dphilipson) [inline packages/accounts/src/msca/plugins/session-key/extension.ts:156]: I couldn't find a way to get this to work without the cast on this line.

**Review by moldy530 — DISMISSED** [review]

**Comment by moldy530** [comment]
oh odd.. we should update the `esm` build steps to use a more recent version of the `module` type, maybe `nodenext` or `es2022`? not sure what the best choice is. I imagine a later `es` version

```
tsc --project tsconfig.build.json --module **es2015** --outDir ./dist/esm --removeComments && echo > ./dist/esm/package.json '{"type":"module"}'
```

**Comment by moldy530** [comment]
ah also, can we change the base branch to v4.x.x

**Comment by dphilipson** [comment]
@moldy530 Done and done!

I changed the `--module` version to es2020, the earliest version mentioned as an alternative in the build error. Let me know if you think this will cause any problems.

**Review by moldy530 — APPROVED** [review]

**Pull request merged**